### ISSUE TITLE
opencv dnn can't load pytorch FasterRCNN exported to onnx detection model

### ISSUE THREAD

**Issue by chenliang2014** [opencv/opencv#21634]
Environment:
- OpenCV => 4.5.5
- Operating System / Platform => Windows10 64 Bit
- pytorch 1.7
- python 3.7

 export pytorch model to onnx code:
`
onnx_model_path = "./fasterrcnn_resnet50_fpn.onnx"
model = torchvision.models.detection.fasterrcnn_resnet50_fpn(pretrained=True)
device = torch.device('cpu')

model.to(device)
dummy_input = torch.randn(1, 3, 608, 608).to(device)
model.eval()
torch.onnx.export(model, dummy_input, onnx_model_path, verbose=False, opset_version=11)
`

In opencv load onnx model code:
`
net =  cv2.dnn.readNetFromONNX(onnx_model_path) 
`

report error:

net =  cv2.dnn.readNetFromONNX(onnx_model_path)
cv2.error: OpenCV(4.5.5) D:\a\opencv-python\opencv-python\opencv\modules\dnn\src\graph_simplifier.cpp:76: error: (-212:Parsing error) Input node with name 1625 not found in function 'cv::dnn::Subgraph::getInputNodeId'

**Comment by alalek** [comment]
We need complete reproducer for investigation including input data.
Please attach `.onnx` model as `.zip` archive to GitHub initial post.

**Comment by chenliang2014** [comment]
@alalek   I can't upload zip file in this page , please download from     
https://drive.google.com/file/d/1yERB0kN1dbWsfh4-Vm4dTPVTYz9AZa8B/view?usp=sharing

**Comment by asmorkalov** [comment]
Diagnostic tool output:
```
./bin/opencv_model_diagnostics -m=/home/alexander/tmp/fasterrcnn_resnet50_fpn.onnx 
[ INFO:0@0.269] global /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/onnx/onnx_importer.cpp (761) populateNet DNN/ONNX: loading ONNX v6 model produced by 'pytorch':1.7. Number of nodes = 1872, initializers = 264, inputs = 1, outputs = 3
[ INFO:0@0.269] global /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/onnx/onnx_importer.cpp (677) parseOperatorSet DNN/ONNX: ONNX opset version = 11
terminate called after throwing an instance of 'cv::Exception'
  what():  OpenCV(4.5.5-dev) /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/graph_simplifier.cpp:76: error: (-212:Parsing error) Input node with name 1625 not found in function 'getInputNodeId'
```

Stack trace:
```
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:51
#1  0x00007ffff4944921 in __GI_abort () at abort.c:79
#2  0x00007ffff4f99957 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#3  0x00007ffff4f9fae6 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#4  0x00007ffff4f9fb21 in std::terminate() () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#5  0x00007ffff4f9fd54 in __cxa_throw () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#6  0x00007ffff570f26f in cv::error (exc=...) at /home/alexander/Projects/OpenCV/opencv-master/modules/core/src/system.cpp:1237
#7  0x00007ffff570f33c in cv::error (_code=-212, _err="Input node with name 1625 not found", 
    _func=0x7ffff7440148 <cv::dnn::Subgraph::getInputNodeId(cv::Ptr<cv::dnn::ImportGraphWrapper> const&, cv::Ptr<cv::dnn::ImportNodeWrapper> const&, int)::__func__> "getInputNodeId", 
    _file=0x7ffff743fff0 "/home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/graph_simplifier.cpp", _line=76) at /home/alexander/Projects/OpenCV/opencv-master/modules/core/src/system.cpp:1249
#8  0x00007ffff6de7972 in cv::dnn::Subgraph::getInputNodeId (net=..., node=..., inpId=1) at /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/graph_simplifier.cpp:76
#9  0x00007ffff6de7d79 in cv::dnn::Subgraph::match (this=0x555555e63d20, net=..., nodeId=744, matchedNodesIds=std::vector of length 0, capacity 16, targetNodesIds=std::vector of length 0, capacity 16)
    at /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/graph_simplifier.cpp:113
#10 0x00007ffff6fd2f57 in cv::dnn::dnn4_v20211220::SoftMaxSubgraphBase::match (this=0x555555e63d20, net=..., nodeId=748, matchedNodesIds=std::vector of length 0, capacity 16, targetNodesIds=std::vector of length 0, capacity 16)
    at /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/onnx/onnx_graph_simplifier.cpp:122
#11 0x00007ffff6de8a37 in cv::dnn::simplifySubgraphs (net=..., patterns=std::vector of length 20, capacity 32 = {...}) at /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/graph_simplifier.cpp:198
#12 0x00007ffff6fdcb07 in cv::dnn::dnn4_v20211220::simplifySubgraphs (net=...) at /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/onnx/onnx_graph_simplifier.cpp:741
#13 0x00007ffff6ff7e5f in cv::dnn::dnn4_v20211220::ONNXImporter::populateNet (this=0x7fffffffd6a0) at /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/onnx/onnx_importer.cpp:765
#14 0x00007ffff6ff1701 in cv::dnn::dnn4_v20211220::ONNXImporter::ONNXImporter (this=0x7fffffffd6a0, net=..., onnxFile=0x5555559e7640 "/home/alexander/tmp/fasterrcnn_resnet50_fpn.onnx")
    at /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/onnx/onnx_importer.cpp:237
#15 0x00007ffff7022772 in cv::dnn::dnn4_v20211220::detail::readNet<cv::dnn::dnn4_v20211220::ONNXImporter, char const*> () at /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/onnx/../dnn_common.hpp:38
#16 0x00007ffff701fcac in cv::dnn::dnn4_v20211220::detail::readNetDiagnostic<cv::dnn::dnn4_v20211220::ONNXImporter, char const*> () at /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/onnx/../dnn_common.hpp:45
#17 0x00007ffff701cdc8 in cv::dnn::dnn4_v20211220::readNetFromONNX (onnxFile="/home/alexander/tmp/fasterrcnn_resnet50_fpn.onnx") at /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/onnx/onnx_importer.cpp:3429
#18 0x00007ffff6da7a84 in cv::dnn::dnn4_v20211220::readNet (_model="/home/alexander/tmp/fasterrcnn_resnet50_fpn.onnx", _config="", _framework="") at /home/alexander/Projects/OpenCV/opencv-master/modules/dnn/src/dnn.cpp:6195
#19 0x0000555555555cb8 in main (argc=2, argv=0x7fffffffdcb8) at /home/alexander/Projects/OpenCV/opencv-master/apps/model-diagnostics/model_diagnostics.cpp:64
```

**Comment by FrancescoMandru** [comment]
I have the same problem with maskrcnn pytorch

**Comment by subika-sri** [comment]
Yeah I too have the same issue with pytorch mask rcnn, I finetuned the model and since I can't able to do batch inference without iterating through data, I needed to convert the model to onnx format and now I am getting this error.

**Comment by ouening** [comment]
I have the same issue:
```python
# original_model = models.detection.retinanet_resnet50_fpn(weights=models.detection.RetinaNet_ResNet50_FPN_Weights.DEFAULT) # do not work
original_model = models.segmentation.fcn_resnet50(weights=models.segmentation.FCN_ResNet50_Weights.DEFAULT) # works
# define the directory for further converted model save
onnx_model_path = "models"
# define the name of further converted model
onnx_model_name = "retinanet.onnx"
# create directory for further converted model
os.makedirs(onnx_model_path, exist_ok=True)
# get full path to the converted model
full_model_path = os.path.join(onnx_model_path, onnx_model_name)
# generate model input to build the graph
generated_input = torch.randn(1, 3, 640, 640)
# model export into ONNX format
torch.onnx.export(
    original_model,
    generated_input,
    full_model_path,
    verbose=False,
    input_names=["input"],
    output_names=["output"],
    opset_version=11
)

# read the transferred network with cv.dnn.readNetFromONNX passing the obtained in the previous step ONNX model into it:
# read converted .onnx model with OpenCV API
opencv_net = cv2.dnn.readNetFromONNX(full_model_path)
```
The official doc in [https://docs.opencv.org/5.x/d2/d58/tutorial_table_of_content_dnn.html](https://docs.opencv.org/5.x/d2/d58/tutorial_table_of_content_dnn.html) only shows the demo on convert pytorch classification and segmentation models to `onnx`. I tried the deetction model in a similar way but failed! Could anybody help?
![image](https://user-images.githubusercontent.com/15122811/203063932-bb8c418b-7e68-4e32-8d3b-99f14af293b7.png)

**Comment by asmorkalov** [comment]
The issue is still reproducible with 4.x branch. @fengyuentau Could you take a look on it?

**Comment by asmorkalov** [comment]
@Abdurrahheem Could you debug it?

**Comment by Abdurrahheem** [comment]
> **Enviroment:**
> 
> > torch version:  1.13.0+cu117
> > torchvision version:  0.14.0+cu117
> > opencv version:  4.6.0
> > numpy version:  1.22.4
> > onnx version:  1.12.0
> 
> I have the same issue:
> 
> ```python
> # original_model = models.detection.retinanet_resnet50_fpn(weights=models.detection.RetinaNet_ResNet50_FPN_Weights.DEFAULT) # do not work
> original_model = models.segmentation.fcn_resnet50(weights=models.segmentation.FCN_ResNet50_Weights.DEFAULT) # works
> # define the directory for further converted model save
> onnx_model_path = "models"
> # define the name of further converted model
> onnx_model_name = "retinanet.onnx"
> # create directory for further converted model
> os.makedirs(onnx_model_path, exist_ok=True)
> # get full path to the converted model
> full_model_path = os.path.join(onnx_model_path, onnx_model_name)
> # generate model input to build the graph
> generated_input = torch.randn(1, 3, 640, 640)
> # model export into ONNX format
> torch.onnx.export(
>     original_model,
>     generated_input,
>     full_model_path,
>     verbose=False,
>     input_names=["input"],
>     output_names=["output"],
>     opset_version=11
> )
> 
> # read the transferred network with cv.dnn.readNetFromONNX passing the obtained in the previous step ONNX model into it:
> # read converted .onnx model with OpenCV API
> opencv_net = cv2.dnn.readNetFromONNX(full_model_path)
> ```
> 
> The official doc in https://docs.opencv.org/5.x/d2/d58/tutorial_table_of_content_dnn.html only shows the demo on convert pytorch classification and segmentation models to `onnx`. I tried the deetction model in a similar way but failed! Could anybody help? ![image](https://user-images.githubusercontent.com/15122811/203063932-bb8c418b-7e68-4e32-8d3b-99f14af293b7.png)



-----------------
REPLY: 



I did not have any problem running this model using the reproduce provided. Specifications are: 

```console
System: Ubuntu 20.04 
OpenCV version: 4.7.0
Torch version: 2.0.0+cpu
Torchvision version: 0.15.1+cpu
Onnx version: 1.11.0
```

**Comment by Abdurrahheem** [comment]
> Environment:
> 
> * OpenCV => 4.5.5
> * Operating System / Platform => Windows10 64 Bit
> * pytorch 1.7
> * python 3.7
> 
> export pytorch model to onnx code: ` onnx_model_path = "./fasterrcnn_resnet50_fpn.onnx" model = torchvision.models.detection.fasterrcnn_resnet50_fpn(pretrained=True) device = torch.device('cpu')
> 
> model.to(device) dummy_input = torch.randn(1, 3, 608, 608).to(device) model.eval() torch.onnx.export(model, dummy_input, onnx_model_path, verbose=False, opset_version=11) `
> 
> In opencv load onnx model code: `net = cv2.dnn.readNetFromONNX(onnx_model_path) `
> 
> report error:
> 
> net = cv2.dnn.readNetFromONNX(onnx_model_path) cv2.error: OpenCV(4.5.5) D:\a\opencv-python\opencv-python\opencv\modules\dnn\src\graph_simplifier.cpp:76: error: (-212:Parsing error) Input node with name 1625 not found in function 'cv::dnn::Subgraph::getInputNodeId'

REPLY: 

I was able to reproduce the error on the following system:

```console
System: Ubuntu 20.04 
Complier: GCC 7.5.0
OpenCV version: 4.7.0 (build from source)
Torch version: 2.0.0+cpu
Torchvision version: 0.15.1+cpu
Onnx version: 1.11.0
```

Working on fixing

**Comment by ukoehler** [comment]
Confirmed for
opencv-python-rolling-4.7.0.20230520
Windows 10
Python 3.10.4

Also tried the usual workaround of applying onnxsim. That just leads to another issue: https://github.com/opencv/opencv/issues/23663

**Comment by Abdurrahheem** [comment]
FasterRCNN will not be support for now. It contains layers such as TopK such "decides" it outputs shape dynamically. Currently such functionality in not supported in OpenCV DNN model. All shapes must be static.
